probe-rs/src/probe/cmsisdap/mod.rs

```diff
@@ -1218,6 +1267,51 @@ impl SwoAccess for CmsisDap {
     }
 }
 
+impl RawJtagIo for CmsisDap {
+    fn state(&self) -> &JtagDriverState {
+        &self.jtag_driver_state
+    }
+    fn state_mut(&mut self) -> &mut JtagDriverState {
+        &mut self.jtag_driver_state
+    }
+    fn shift_bit(&mut self, tms: bool, tdi: bool, capture: bool) -> Result<(), DebugProbeError> {
+        self.jtag_driver_state.state.update(tms);
+        if let Some(sequence) = self.jtag_sequences.last_mut() {
+            if sequence.append(tms, tdi, capture).is_ok() {
+                return Ok(());
+            }
+        }
+        self.jtag_sequences.push(JtagSequence::new(
+            1,
+            capture,
+            tms,
+            [u8::from(tdi), 0, 0, 0, 0, 0, 0, 0],
+        )?);
+        Ok(())
+    }
+    fn read_captured_bits(&mut self) -> Result<BitVec<u8, Lsb0>, DebugProbeError> {
+        let mut capture = BitVec::<u8, Lsb0>::new();
+        // We can transfer up to 255 sequences in one dap jtag-sequence command. But in some riscv chip there is some bug, thus I set up to 3 sequence in once jtag-sequence command transfer.
+        for sequence_slice in self.jtag_sequences.clone().chunks(3 as usize) {
```

Comment:
Why do you clone here? If you need to own the sequences, use `into_iter` or `drain`, the latter allows you to remove the `clear()` calls later on.